lib/google/gax/paged_enumerable.rb

```diff
@@ -0,0 +1,200 @@
+# in the documentation and/or other materials provided with the
+# distribution.
+#     * Neither the name of Google Inc. nor the names of its
+# contributors may be used to endorse or promote products derived from
+# this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+module Google
+  module Gax
+    # A class to provide the Enumerable interface for page-streaming method.
+    # PagedEnumerable assumes that the API call returns a message for a page
+    # which holds a list of resources and the token to the next page.
+    #
+    # PagedEnumerable provides the enumerations over the resource data,
+    # and also provides the enumerations over the pages themselves.
+    #
+    # @example normal iteration over resources.
+    #   paged_enumerable.each { |resource| puts resource }
+    #
+    # @example per-page iteration.
+    #   paged_enumerable.each_page { |page| puts page }
+    #
+    # @example Enumerable over pages.
+    #   paged_enumerable.each_page do |page|
+    #     page.each { |resource| puts resource }
+    #   end
+    #
+    # @example more exact operations over pages.
+    #   while some_condition()
+    #     page = paged_enumerable.page
+    #     do_something(page)
+    #     break if paged_enumerable.next_page?
+    #     paged_enumerable.next_page
+    #   end
+    #
+    class PagedEnumerable
+      include Enumerable
+
+      # @attribute [r] page
+      #   @return [Page] The current page object.
+      attr_reader :page
+
+      # @param request_page_token_field [String]
+      #   The name of the field in request which will have the page token.
+      # @param response_page_token_field [String]
+      #   The name of the field in the response which holds the next page token.
+      # @param resource_field [String]
+      #   The name of the field in the response which holds the resources.
+      def initialize(api_call, request, response, options: nil,
+                     resource_field: nil)
+        @api_call = api_call
+        @request = request
+        @response = response
+        @options = options
+        @resource_field = resource_field || determine_resource_field(response)
```

Comment:
Do we need this default vs raising if the `resource_field` is not specified? (Also need to update the method docs)

Reply:
I could go either way. We will probably be adding the logic to determine the first repeated field in the generator. But, since paging is supposed to be implemented by convention I had also thought that the convention could be followed in the runtime if the field name wasn't provided.

I had also wondered if we want to check the `request` object for `page_token` and `page_size`, and the `response` object for `next_page_token` and a repeated field, per the convention. That could possibly save the need for passing a resource_field, and ensure that anyone using GAX outside of the new generator are using it correctly.

What do you think?

Reply:
That sounds fine to me. In that case, I think we may want to make tighten the check in `#determine_resource_field` to ensure that there is exactly one `:repeated` field when it's inferred.

Reply:
Ah, that was part of what I was trying to figure out earlier this week. The unofficial word is that we take the first repeated field, but are you saying that there can only be one repeated field in a paginated response? Is that the official convention?

Reply:
The "official" convention is IMHO a bit convoluted:

> If the response message has more than one non-primitive repeated field, the first one (in order of appearance in the file and field number) is used. If the first field by order of appearance in the message and the first field by field number do not match, code generators that implement automatic pagination should fail with an error.

But, we don't really have to implement this to the letter. If we go with this as-is then we also need to check the field number to be consistent with the convention. However, I think we should simply require the user to explicitly state the field that they want to use if there's more than one. That's easier to explain than the field number check and it will make more sense if gax is used outside of the generator.

Reply:
Oh, are primitive fields allowed to be paginated then?

Reply:
No, primitive fields will not be paged by convention. We could allow primitives to be paged in gax if we really wanted to, but we'll never use it in the generator, so I don't think it's worth worrying about unless we get an external ask for it.

Reply:
K, I have added a commit that attempts to implement these rules. There are no tests yet, but they will come later. There are a few more changes I would like to make before backfilling the tests. But they are coming.